oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtDecoderBuilder.java

```diff
@@ -0,0 +1,223 @@
+		Set<SignatureAlgorithm> jwkAlgorithms = getDefaultAlgorithms();
+		try {
+			jwkAlgorithms.addAll(fetchSignatureVerificationAlgorithms(jwkSource));
+		} catch (Exception ex) {
+			log.error("Error fetching Signature Verification algorithms");
+		}
+		return convertToJwsAlgorithms(jwkAlgorithms);
+	}
+
+	/**
+	 * Fetches {@link SignatureAlgorithm}s based on the configured {@link ReactiveJWKSource}s keys.
+	 * @param jwkSource
+	 * @return A set of {@link JWSAlgorithm}s to be used for JWT signature verification.
+	 */
+	protected Set<JWSAlgorithm> getSignatureAlgorithms(ReactiveJWKSource jwkSource) {
+		Set<SignatureAlgorithm> jwkAlgorithms = getDefaultAlgorithms();
+		try {
+			jwkAlgorithms.addAll(fetchSignatureVerificationAlgorithms(jwkSource));
+		} catch (Exception ex) {
+			log.error("Error fetching Signature Verification algorithms");
+		}
+		return convertToJwsAlgorithms(jwkAlgorithms);
+	}
+
+	/**
+	 * Retains the original functionality for adding {@link SignatureAlgorithm#RS256} as a default algorithm if none are provided.
+	 * @return A set of default {@link SignatureAlgorithm}s
+	 */
+	private Set<SignatureAlgorithm> getDefaultAlgorithms() {
+		Set<SignatureAlgorithm> jwkAlgorithms = new HashSet<>();
+		if (this.signatureAlgorithms.isEmpty()) {
+			jwkAlgorithms.add(SignatureAlgorithm.RS256);
+		} else {
+			jwkAlgorithms.addAll(this.signatureAlgorithms);
+		}
+		return jwkAlgorithms;
+	}
+
+	private Set<JWSAlgorithm> convertToJwsAlgorithms(Set<SignatureAlgorithm> algorithms) {
+		return algorithms.stream()
+				.map(algorithm -> JWSAlgorithm.parse(algorithm.getName()))
+				.collect(Collectors.toSet());
+	}
+
+	/**
+	 * Given a valid {@link JWKSource}, fetches, and parses out the algorithms of available JWKs.
+	 * @param jwkSource
+	 * @return A set of {@link SignatureAlgorithm} instances that may be used to validate a JWT (JWS).
+	 */
+	private Set<SignatureAlgorithm> fetchSignatureVerificationAlgorithms(JWKSource<SecurityContext> jwkSource) {
+		return fetchSignatureVerificationAlgorithms(fetchSignatureVerificationJwks(jwkSource));
+	}
+
+	/**
+	 * Given a valid {@link ReactiveJWKSource}, fetches, and parses out the algorithms of available JWKs.
+	 * @param jwkSource
+	 * @return A set of {@link SignatureAlgorithm} instances that may be used to validate a JWT (JWS).
+	 */
+	private Set<SignatureAlgorithm> fetchSignatureVerificationAlgorithms(ReactiveJWKSource jwkSource) {
+		return fetchSignatureVerificationAlgorithms(fetchSignatureVerificationJwks(jwkSource));
+	}
+
+	/**
+	 * Converts a list of {@link JWK}s into a set of {@link SignatureAlgorithm}s.
+	 * @param jwks
+	 * @return A set of {@link SignatureAlgorithm} instances that may be used to validate a JWT (JWS).
+	 */
+	private Set<SignatureAlgorithm> fetchSignatureVerificationAlgorithms(List<JWK> jwks) {
+		if (jwks == null) {
+			return Collections.emptySet();
+		}
+		return jwks.stream().map(jwk -> {
```

Comment:
The Stream API is a bit too slow and hard on the garbage collector to be used in Spring Security - will you please change this to a for loop?

Reply:
Yes that is no problem at all.